src/OpenTelemetry/Trace/TracerProviderSdk.cs

```diff
@@ -71,7 +83,7 @@ internal sealed class TracerProviderSdk : TracerProvider
                     if (this.supportLegacyActivity && string.IsNullOrEmpty(activity.Source.Name))
```

Comment:
Given `supportLegacyActivity` is something we can determine ahead of time, consider extracting the logic and create the corresponding delegate with hard-wired logic, so we don't have to check it in every Activity start/stop.

```csharp

if (this.supportLegacyActivity)
{
    listener.ActivityStarted = (activity) =>
    {
        if (string.IsNullOrEmpty(activity.Source.Name))
        {
            // legacy logic
            // similar here, we already know whether wildcard mode should be used or not, we can hard-wire the logic
            return;
        }
        // new logic
    }
}
else
{
    listener.ActivityStarted = (activity) =>
    {
        // new logic
    }
}

Reply:
Similar approach has been used here to optimize the hot path:
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/bd189aa654948e79203d66d4b6d33fb93557d7e9/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L185
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/bd189aa654948e79203d66d4b6d33fb93557d7e9/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L206

Reply:
That makes sense and thanks for the tip. I've had a go at an initial optimization, let me know if this is what you expected.

Benchmark results (the top table is pre-optimisation, bottom is post), they're pretty marginal gains:

![image](https://user-images.githubusercontent.com/8128694/135770529-1a32098a-8b8a-4f5f-b2e8-a004bf6d684f.png)

Reply:
👍  :shipit: